### DISCUSSION TITLE
Strange checker-board output by display

### DISCUSSION THREAD

**Discussion by porphyry5** [ImageMagick/ImageMagick#2319] [Help] [answered]
Not sure if this is a feature or a bug, I have a vague memory that display produces this checker-board pattern in response to some file condition it objects to. But there is no error message here to indicate display's objection.

xv.jpg and disp.jpg are views of the same file by xv and display. xv is showing what the file does contain.

The file is produced by a bash function I'm developing, to show a montage of various -lat operations on the same file.

display was showing the file correctly until I changed this command
`convert x.tiff -lat $c$q$c$i%  -annotate +15+$d "-lat $c$q$c$i%" y.tiff`
to 
`convert x.tiff -annotate +15+$d "-lat $c$q$c$i%" -lat $c$q$c$i%  y.tiff`
so that the annotation text would always have a contrasting background.

Could someone kindly advise me if there is something I need to do to prevent this outcome. I prefer to use display to show the finished product, as it does not shrink to fit the screen like xv.
![disp](https://user-images.githubusercontent.com/10132779/88463726-ba420c00-ce69-11ea-8bd0-dd5dd20362c0.jpg)
![xv](https://user-images.githubusercontent.com/10132779/88463728-be6e2980-ce69-11ea-9d99-d0d5223d3836.jpg)

**Comment by fmw42** [comment, marked as answer]
The display checkerboard likely shows where the image is transparent.

**Reply by porphyry5** [reply]
If that is so, why would including a one-line annotation at the bottom of each image before the -lat command, make the entire image transparent?

**Reply by fmw42** [reply]
Post your tiff file so others can test with it. Also what is your IM version and platform?  Why do you have two "-lat" commands one right after the other in your command?

**Reply by porphyry5** [reply]
~ $ identify -version
Version: ImageMagick 6.9.7-4 Q16 x86_64 20170114 http://www.imagemagick.org
Copyright: © 1999-2017 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP 
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff wmf x xml zlib

running on
~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.4 LTS
Release:        18.04
Codename:       bionic

The file concerned is a .tiff, which cannot be uploaded here. But the same result happens with a jpg, so I have provided that instead.
![109](https://user-images.githubusercontent.com/10132779/88566866-7e36b480-cfeb-11ea-8025-6e6e38ce0b06.jpg)

Also following is text of bash function producing the file
`lc(){ # variations on a theme by lat, lc file "1st inc last" (def -11 2 1)
  a=($(identify $1))
  b=(${a[2]/x/ })
  w=160
  h=160
  x=$(((${b[0]}-$w)/2))
  y=$(((${b[1]}-$h)/2))
  convert $1 -crop "$w"x$h+$x+$y -bordercolor white -border 2x20 +repage x.tiff
  mogrify -crop +0+20 +repage x.tiff
  c=5
  let d=$h+15
  e="${2:--11 2 1}"
  q='x'
  for ((j=1; j<5; j++)); do
    convert x.tiff -annotate +15+$d "original" z.tiff
    let c=c+5
    for i in $(seq -s ' ' $e); do
      if [ $i -gt 0 ]; then i=+$i; fi
      convert x.tiff -annotate +15+$d "-lat $c$q$c$i%" -lat $c$q$c$i%  y.tiff
      convert z.tiff y.tiff +append z.tiff
    done
    if (( j > 1 )); then
      convert q.tiff z.tiff -append q.tiff
    else
      mv z.tiff q.tiff
    fi
  done
  display q.tiff
  xv q.tiff
}
`

**Reply by porphyry5** [reply]
Sorry about the loss of format on the function, and I can't find any way to edit after the fact

**Reply by fmw42** [reply]
Too hard to read. It would have been better if you showed an example with actual number in place of your variables and no loops.

I would suggest that you put in +write tmpX.png statements after each operation and find where it is deviating from what you expect.

**Reply by porphyry5** [reply]
Here are 4 commands with numbers that will produce the effect

`convert 109.tiff -crop 160x160+125+180 -bordercolor white -border 2x20 +repage x.tiff
mogrify -crop +0+20 +repage x.tiff
convert x.tiff -annotate +15+175 "-lat 5x5-3%" -lat 5x5-3%  y.tiff
display y.tiff
`

As I stated in the first post, it is the final convert that **deviates,** and only because I reversed the order of the -lat and -annotate commands, when the -lat is done before the -annotate the result is normal

**Reply by porphyry5** [reply]
Never mind, I'll post a file of those 4 commands for you, test.txt
[test.txt](https://github.com/ImageMagick/ImageMagick/files/4990592/test.txt)

**Reply by fmw42** [reply]
Those two commands do quite different things. You have annotate between two -lat in the first command. So something is added to your image before the second lat. In the second command you annotate and then do two -lats. Can you provide your 109.tiff image, so I can test both commands.

**Reply by porphyry5** [reply]
No, the -annotate states what the -lat command is **within quotes**, there is only one actual -lat command

**Reply by fmw42** [reply]
Sorry, I misread that. But in your text.txt file, your second command processes the image resulting from the first two commands. But in your original post, your two commands start with the same image.  So I am confused by your test.txt commands. Can you provide just two convert commands for me to compare? Sorry, for the extra work. But I need a direct comparison.

But I will point out that order is important in processing. Not all operations are commutable.

**Reply by porphyry5** [reply]
![109](https://user-images.githubusercontent.com/10132779/88712010-ea89e480-d0cd-11ea-8f58-933e25994dab.jpg)
[test.txt](https://github.com/ImageMagick/ImageMagick/files/4990658/test.txt)

Your thing here won't upload tiffs, so here it is as jpg, and I've altered test.txt to operate on jpg instead.

BTW, is the problem with pasting commands into this editor that it only recognizes windows style line-breaks?

**Reply by fmw42** [reply]
I do not understand. I have no problem with Mac/Unix style command line breaks. You should paste your code and then highlight your text and use the <> button to format it so that the line breaks are preserved.

Sorry, your test.txt file is confusing to me. It does not seem to have two simple commands for me to compare. Please, I need the one convert command that works and the second convert command that you modified the order of the -annotate and -lat that fails.

**Reply by porphyry5** [reply]
Just run the final convert on 109.jpg,

**Reply by fmw42** [reply]
-annotate adds an (opaque i.e. white) alpha channel to your processing. -lat process the alpha channel as well as your underlying image. The processed alpha channel then turns to black so that it makes the image transparent. Turn off the alpha channel after the -annotate and your commands will work as you expect. So

`convert 109.jpg -lat 5x5-3% -annotate +15+175 "-lat 5x5-3%" -alpha off x.tiff`

and

`convert 109.jpg -annotate +15+175 "-lat 5x5-3%" -alpha off -lat 5x5-3%  y.tiff`

**Reply by porphyry5** [reply]
Thank you so much for this info, and also for how to paste code in previous response. I always use the path of least apparent effort, so would select code tags first, then paste between them.

One final request, if I may. Could you recommend a text giving non-technical definition/discussion of the theory of computer imaging, as I have zero familiarity with it, and its now apparent I need to know this. I have only the haziest notion of transparencies, alpha or other greek channels, and many of ImageMagick's options are total unknowns to me.

**Reply by porphyry5** [reply]
Sorry, also, how do I mark the question as "Answered"?

**Reply by fmw42** [reply]
Click the green button in the top right of the answer.  See https://www.thriftbooks.com/w/digital-image-processing_rafael-c-gonzalez_richard-e-woods/266327/item/7913824/?mkwid=%7cdc&pcrid=448918537671&pkw=&pmt=&slid=&plc=&pgrid=107461511234&ptaid=pla-926306231989&gclid=EAIaIQobChMIscykjZbz6gIV4D6tBh214QK_EAQYASABEgLGXvD_BwE#isbn=013168728X&idiq=791382 and https://www.goodreads.com/book/show/658971.Fundamentals_of_Digital_Image_Processing and https://www.goodreads.com/book/show/21382100-introduction-to-image-processing-and-analysis

**Reply by fmw42** [reply]
Also for ImageMagick, see 

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
